### PR TITLE
preliminary support for tf.cond and tf.case

### PR DESCRIPTION
**Pull request by lucienwang1009**
- Preliminarily support nest tf.cond and tf.case.
- Put construct_graph_from_nodes into utils.
- Amend delete_unused_nodes to delete unused nodes in subgraphs of nodes recursively. 
- Add input_shape_at used in shape_inference.
- Fix Identity conversion bug with replace_all_inputs.
- Fix run_pretrained_models bug related to force_input_shape.

| Tests                      | Status                     |
|----------------------------|----------------------------|
| unittests                  | success                    |
| test_custom_rnncell        | success                    |
| test_loop                  | success                    |
| test_run_pretrained_models | success                    |
| internal-cnn               | 1 failed is an known issue |
| internal-rnn               | 1 failed is an known issue |

### CODE REVIEW

File: tf2onnx/rewriter/cond_rewriter.py [hunk 1]
```diff
@@ -105,8 +104,43 @@ def run(self):
...
+        for node in nodes:
+            for inp in node.input:
+                input_node = self.g.get_node_by_output(inp)
+                if self._is_switch(input_node):
+                    if inp == input_node.output[0]:
+                        if branch == BranchType.TRUE:
+                            raise ValueError(
+                                "true and false graph intersect at {}".format(node.name)
+                            )
+                        branch = BranchType.FALSE
```
Comment (pengwa) [inline tf2onnx/rewriter/cond_rewriter.py]: should you break here?

File: tf2onnx/rewriter/cond_rewriter.py [hunk 2]
```diff
@@ -105,8 +104,43 @@ def run(self):
...
+                    else:
+                        if branch == BranchType.FALSE:
+                            raise ValueError(
+                                "true and false graph intersect at {}".format(node.name)
+                            )
+                        branch = BranchType.TRUE
```
Comment (pengwa) [inline tf2onnx/rewriter/cond_rewriter.py]: same here

File: tf2onnx/rewriter/cond_rewriter.py [hunk 3]
```diff
@@ -139,150 +173,112 @@ def _cut_off_connection(self, cond_context):
...
+            [merge_input_1],
+            stop_at_switch
+        )
+        branch_nodes_2 = self.g.extract_sub_graph_nodes(
+            [merge_input_2],
+            stop_at_switch
+        )
+        branch_type_1 = self._branch_type(merge_input_1, branch_nodes_1)
+        branch_type_2 = self._branch_type(merge_input_2, branch_nodes_2)
+        # handle the case one branch only contains a const node
```
Comment (pengwa) [inline tf2onnx/rewriter/cond_rewriter.py]: what's the comment for?

File: tf2onnx/rewriter/cond_rewriter.py [hunk 4]
```diff
@@ -139,150 +173,112 @@ def _cut_off_connection(self, cond_context):
...
+        if branch_type_1 == BranchType.TRUE:
+            true_branch_nodes = branch_nodes_1
+            true_output = merge_input_1
+            false_branch_nodes = branch_nodes_2
+            false_output = merge_input_2
+        else:
+            true_branch_nodes = branch_nodes_2
+            true_output = merge_input_2
+            false_branch_nodes = branch_nodes_1
+            false_output = merge_input_1
```
Comment (pengwa) [inline tf2onnx/rewriter/cond_rewriter.py]: how would you handle cases where one branch type is unknown, another one is true or false. Is this is supported? If not , throw exception please

File: tf2onnx/rewriter/cond_rewriter.py [hunk 5]
```diff
@@ -105,8 +104,43 @@ def run(self):
 
         return self.g.get_nodes()
 
+    def _branch_type(self, branch_output, nodes):
```
Comment (pengwa) [inline tf2onnx/rewriter/cond_rewriter.py]: put this after _trace_back_from_one_merge pls.

File: tf2onnx/rewriter/cond_rewriter.py [hunk 6]
```diff
@@ -139,150 +173,112 @@ def _cut_off_connection(self, cond_context):
...
-                    else:
-                        true_graph_context.output.add(i)
+            true_branch_nodes, true_output, false_branch_nodes, false_output, switchs = \
+                self._trace_back_from_one_merge(merge_node)
+            true_branch_context.nodes |= set(true_branch_nodes)
+            true_branch_context.output.append(true_output)
+            false_branch_context.nodes |= set(false_branch_nodes)
+            false_branch_context.output.append(false_output)
+            total_switchs |= switchs
+        log.debug("=================true body graph===============")
```
Comment (pengwa) [inline tf2onnx/rewriter/cond_rewriter.py]: remove those debug info before check in

File: tf2onnx/rewriter/cond_rewriter.py [hunk 7]
```diff
@@ -139,150 +173,112 @@ def _cut_off_connection(self, cond_context):
         nodes_to_add.append(cond_context.if_node)
```
Comment (pengwa) [inline tf2onnx/rewriter/cond_rewriter.py]: put node creation along with _set_branch_graph logic into a new function.

**Comment by nbcsm** [comment]
@lucienwang1009, find a few typo: switchs => swtiches, branchs => branches. 
could you help to get them fixed soemtime? no hurry.

**Comment by lucienwang1009** [comment]
@nbcsm sure will fix it in next PR.

**Pull request merged**